towerlib/towerlib.py

```diff
@@ -1997,24 +2011,579 @@ def schedules(self):
+        credential id in the config file .
+        This function will only be called if no credential id is given in the config file. Anyway, if no credential is
+        given by config file and no credential id is found with this function, the script will continue and create
+        projects, because projects can be created in Ansible Tower without credentials (but in that case, the scm
+        update job will fail.)
+
+        Args:
+            scm_url: the scm_url of the corresponding project.
+
+        Returns:
+            id of a credential if corresponding project exists, else None.
+
+        """
+        scm_credential_type_ids = []
+        credential_types = list(self.credential_types)
+        for credential_type in credential_types:
+            if credential_type.kind == 'scm':
+                scm_credential_type_ids.append(credential_type.id)
+        credential_id = None
+        projects = list(self.projects)
+        for scm_credential_type_id in scm_credential_type_ids:
+            for project in projects:
+                if project.scm_url == scm_url:
+                    self._logger.debug("Project found with the scm credential.")
+                    if project.credential.credential_type.id == scm_credential_type_id:
+                        credential_id = project.credential.id
+                        if credential_id:
+                            self._logger.debug("Credential ID found from the existing project '{}' and the id is '{}'"
+                                               .format(project.name, credential_id))
+                            return credential_id
+        return credential_id
+
+    def get_projects_by_scm_url_and_branch(self, scm_url, scm_branch):
+        """Get the ansible tower projects with given scm_url and branch name.
+
+        Args:
+            scm_url: given scm url.
+            scm_branch: given scm branch.
+
+        Returns:
+            list: list of matching projects
+
+        """
+        matching_projects = []
+        projects = list(self.projects)
+        for project in projects:
+            if project.scm_type == 'git':
+                if project.scm_url == scm_url and project.scm_branch == scm_branch:
+                    matching_projects.append(project)
+        return matching_projects
+
+    def get_projects_by_scm_url(self, scm_url):
+        """Get the ansible tower projects with given scm_url.
+
+        Args:
+            scm_url: the scm url of the production branch project.
+
+        Returns:
+            The id of the production branch project.
+
+        """
+        matching_projects = []
+        projects = list(self.projects)
+        for project in projects:
+            if project.scm_type == 'git':
+                if project.scm_url == scm_url:
+                    matching_projects.append(project)
+        return matching_projects
+
+    def get_playbooks_by_scm_url_and_branch(self, scm_url, scm_branch):
```

Comment:
I think that all of these can be implemented client side as a filter on the projects and should not be part of the library.